---
title: "azimuth_labeling"
author: "Osman Sharifi"
date: '2022-06-23'
output: html_document
---


```{r loadlibraries}
library(Seurat)
library(Azimuth)
library(SeuratData)
library(patchwork)
library(scCustomize)
library(dplyr)
library(ggplot2)
```

## Map human cortex scRNA-seq datasets from the allen atlas to the RETT corticies

```{r cortex_load, results='hide'}
# Install the PBMC systematic comparative analyis (pmbcsca) dataset
InstallData("humancortexref")
# returns a Seurat object named hrett_cortex
LoadData("humancortexref")
#The RunAzimuth function can take a Seurat object as input
load("/Users/osman/Desktop/LaSalle_lab/Human_rett_data/NovaSeq_data/05_Human_rett_corrected2.RData")
human_rett_cort_filt <- RunAzimuth(experiment.aggregate, reference = "humancortexref")
#Normalize data
human_rett_cort_filt <- NormalizeData(human_rett_cort_filt)
human_rett_cort_filt <- ScaleData(human_rett_cort_filt)
Idents(human_rett_cort_filt) <- "predicted.subclass"
```
```{r load seurat object directly}
load("/Users/osman/Documents/GitHub/snRNA-seq-pipeline/scripts/07_human_cell_labeling/human_rett_cort_filt.RData")
# Rename classes.
human_rett_cort_filt <- RenameIdents(object = human_rett_cort_filt, `L2/3 IT` = "L2_3_IT", `L6 CT` = "L6", `L6 IT` = "L6", `L6b` = "L6", `L6 IT Car3` = "L6", `Sst Chodl` = "Sst", `L5 IT` = "L5", `L5 ET` = "L5", `Micro-PVM` = "Micro", `VLMC` = "Peri", `L5/6 NP` = "L5_6_NP")
levels(human_rett_cort_filt) <- c("L2_3_IT", "L5",  "L5_6_NP", "L6","Pvalb", "Vip", "Sst","Sncg","Lamp5","Peri", "Endo", "Oligo", "OPC", "Astro","Micro")
```
We will add metadata so we can parse by sex, age and disease later.
```{r add metadata}
sample_names <- c("1136_CTRL","1420_RTT","1711_CTRL","1815_RTT","1846_CTRL","3381_RTT","4321_RTT","4591_CTRL","4687_RTT",  "4852_RTT","5020_RTT","5125_CTRL","5161_CTRL","662_CTRL","812_CTRL")

#parse by condition
human_rett_cort_filt$Condition <- plyr::mapvalues(
  x = human_rett_cort_filt$orig.ident, 
  from = c(sample_names), 
  to = c("WT", "RTT", "WT", "RTT", "WT", "RTT", "RTT", "WT", "RTT", "RTT", "RTT","WT", "WT","WT", "WT" )
)
#parse by age
human_rett_cort_filt$Age <- plyr::mapvalues(
  x = human_rett_cort_filt$orig.ident, 
  from = c(sample_names), 
  to = c(34, 21.1, 27, 18, 20.6, 22, 23, 16, 8, 19.8, 24,24.4, 10.7,12, 18.4)
)

```

We can now visualize the outputs of Azimuth. Note that all cells are labeled with high-resolution annotations, and are projected into a harmonized space despite being analyzed from a wide variety of technologies.

```{r cortex_dimplot, fig.width=12, fig.height=5}
human_rett_cort_filt <- RenameIdents(
  object = human_rett_cort_filt,
  'Exc L2 LINC00507 GLRA3' = 'L2_3_IT', 
  'Exc L2-3 RORB CCDC68' = 'L2_3_IT', 
  'Exc L2-3 RORB PTPN3' = 'L2_3_IT', 
  'Exc L3 LAMP5 CARM1P1' = 'L2_3_IT', 
  'Exc L3 RORB OTOGL' = 'L2_3_IT', 
  'Exc L3 THEMIS ENPEP' = 'L2_3_IT', 
  'Exc L3-5 RORB LINC01202' = 'L4', 
  'Exc L3-5 RORB LNX2' = 'L4', 
  'Exc L3-5 RORB RPRM' = 'L4', 
  'Exc L5 FEZF2 CSN1S1' = 'L5',
  'Exc L5 THEMIS SLC22A18' = 'L5',
  'Exc L5-6 FEZF2 C9orf135-AS1' = 'L5',
  'Exc L5-6 FEZF2 IFNG-AS1' = 'L5',
  'Exc L5-6 THEMIS SMYD1' = 'L6',     
  'Exc L5-6 THEMIS TNFAIP6' = 'L6',
  'Exc L6 FEZF2 KLK7' = 'L6',
  'Exc L6 FEZF2 PROKR2' = 'L6',     
  'Exc L6 THEMIS LINC00343' = 'L6',
  'Inh L1 LAMP5 BMP2' = 'LAMP5',
  'Inh L1 LAMP5 PVRL2' = 'LAMP5',
  'Inh L1-6 LAMP5 AARD' = 'LAMP5',
  'Inh L1-6 LAMP5 CA1' = 'LAMP5',
  'Inh L1-6 LAMP5 NES' = 'LAMP5',
  'Inh L5-6 LAMP5 CRABP1' = 'LAMP5',
  'Inh L1 PAX6 CHRFAM7A' = 'LAMP5',
  'Inh L1 PVALB SST ASIC4' = 'PVALB',
  'Inh L1-6 PVALB COL15A1' = 'PVALB',
  'Inh L2-5 PVALB RPH3AL' = 'PVALB',
  'Inh L3 PVALB SAMD13' = 'PVALB',
  'Inh L5-6 PVALB GAPDHP60' = 'PVALB',
  'Inh L5-6 PVALB MEPE' = 'PVALB', 
  'Inh L5-6 PVALB SST CRHR2' = 'PVALB',
  'Inh L1 SST DEFB108B' = 'SST',
  'Inh L1 SST P4HA3' = 'SST', 
  'Inh L1-2 SST CLIC6' = 'SST',
  'Inh L1-2 SST PRRT4' = 'SST',
  'Inh L1-6 SST NPY' = 'SST',
  'Inh L1-3 SST FAM20A' = 'SST',
  'Inh L2-3 SST NMU' = 'SST', 
  'Inh L3-5 SST GGTLC3' = 'SST',
  'Inh L5 SST RPL35AP11' = 'SST',
  'Inh L5-6 SST BEAN1' = 'SST',
  'Inh L5-6 SST C4orf26' = 'SST',
  'Inh L5-6 SST ISX' = 'SST', 
  'Inh L5-6 SST KLHL1' = 'SST',
  'Inh L6 SST TH' = 'SST',
  'Inh L1-3 VIP CBLN1' = 'VIP', 
  'Inh L1-3 VIP CHRNA2' = 'VIP',          
  'Inh L1-3 VIP FNDC1' = 'VIP',
  'Inh L1-3 VIP HSPB6' = 'VIP',
  'Inh L1-6 VIP SLC7A6OS' = 'VIP',
  'Inh L2 VIP SLC6A16' = 'VIP', 
  'Inh L2-5 VIP SOX11' = 'VIP',
  'Inh L1-5 VIP LINC01013' = 'VIP', 
  'Inh L1-5 VIP SMOC1' = 'VIP',
  'Inh L3-5 VIP IGDCC3' = 'VIP',
  'Inh L3-5 VIP TAC3' = 'VIP',
  'Inh L3-6 VIP ZIM2-AS1' = 'VIP',
  'Inh L5-6 VIP COL4A3' = 'VIP',
  'Astro L1 FGFR3 SERPINI2' = 'ASTRO', 
  'Astro L1-6 FGFR3 AQP1' = 'ASTRO', 
  'Astro L1-6 FGFR3 PLCG1' = 'ASTRO',  
  'Oligo L2-6 OPALIN FTH1P3' = 'OLIGO', 
  'Oligo L3-6 OPALIN ENPP6'= 'OLIGO', 
  'Oligo L3-6 OPALIN LRP4-AS1' = 'OLIGO', 
  'OPC L1-6 PDGFRA COL20A1' = 'OLIGO',
  'OPC L1-6 PDGFRA COL20A1' = 'OLIGO',
  'Endo L2-5 NOSTRIN SRGN' = 'ENDO',
  'Micro L1-6 TYROBP CD74' = 'MICRO',
  'VLMC L1-5 PDGFRA COLEC12' = 'PERI'
)
DimPlot_scCustom(seurat_object = human_rett_cort_filt, label = FALSE)
```
Let's look at MECP2 expression
```{r make umap groupings}
use.pcs <- 1:18
human_rett_cort_filt <- RunUMAP(
  object = human_rett_cort_filt,
  dims = use.pcs)

p1 <- FeaturePlot_scCustom(seurat_object = human_rett_cort_filt, features = 'MECP2', split.by = "Condition")
p2 <- DimPlot_scCustom(seurat_object = human_rett_cort_filt, label = FALSE)
wrap_plots(p1/p2)
ggsave("MECP2_UMAP.pdf",
       device = NULL,
       height = 8.5,
       width = 12)

VlnPlot(human_rett_cort_filt, features = "MECP2", split.by = "Condition")
ggsave("MECP2_vlnplot.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
```
UMAP plots of data 
```{r UMAP plots of data}
DimPlot_scCustom(seurat_object = human_rett_cort_filt)
ggsave("celltypes_UMAP.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
DimPlot_scCustom(seurat_object = human_rett_cort_filt, group.by = "predicted.class") +ggtitle("High level labels")
ggsave("celltypes_highlevel_UMAP.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
DimPlot_scCustom(seurat_object = human_rett_cort_filt, group.by = "predicted.cluster") +ggtitle("Low level labels")+ theme(legend.position="bottom")
ggsave("celltypes_lowlevel_UMAP.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
DimPlot_scCustom(seurat_object = human_rett_cort_filt, group.by = "Condition")
ggsave("Conditon_UMAP.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
DimPlot_scCustom(seurat_object = human_rett_cort_filt, group.by = "Samples")
ggsave("samples_UMAP.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
FeaturePlot_scCustom(seurat_object = human_rett_cort_filt, features = "percent_mito")
ggsave("mito_UMAP.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
```
Find the marker genes for celltypes
```{r marker genes}
# Run basic FindAllMarkers DE test
all_markers <- FindAllMarkers(object = human_rett_cort_filt, verbose = F)

all_markers <- all_markers %>%
    group_by(cluster) %>%
    arrange(desc(avg_log2FC)) %>%
    arrange(cluster)
# run piped with Seurat command
all_markers_pct <- all_markers %>%
    Add_Pct_Diff()
all_markers_pct <- all_markers_pct %>%
    group_by(cluster) %>%
    arrange(desc(avg_log2FC)) %>%
    arrange(cluster)
all_markers_pct <- Add_Pct_Diff(marker_dataframe = all_markers)
# By default it returns a named vector of genes (each gene named for cluster it was associated
# with)
top_5 <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 5, rank_by = "avg_log2FC")
top5_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 3, named_vector = FALSE,
    make_unique = TRUE)
head(top_5, 10)

Stacked_VlnPlot(seurat_object = human_rett_cort_filt, features = top_5, x_lab_rotate = TRUE)
Clustered_DotPlot(seurat_object = human_rett_cort_filt, features = top5_markers)

DoHeatmap(
    object = human_rett_cort_filt, 
    features = top_5
) + NoLegend()

top5 <- markers_all_single %>% group_by(cluster) %>% top_n(5, avg_log2FC)
dim(top5)
polychrome_pal <- DiscretePalette_scCustomize(num_colors = 36, palette = "polychrome")
DoHeatmap(
    object = human_rett_cort_filt, 
    features = allen_markers,
    group.colors = polychrome_pal,
    group.bar = TRUE,
    raster = FALSE,
    slot = 'data',
) + scale_fill_gradientn(colors = c("blue", "red"))
ggplot2::ggsave("allen_markers_heatmap.pdf",
                device = NULL,
                height = 8.5,
                width = 12,
                dpi = 500)
DoHeatmap(
    object = human_rett_cort_filt, 
    features = top5_markers,
    group.colors = polychrome_pal,
    group.bar = TRUE,
    raster = FALSE,
    slot = 'data',
) + scale_fill_gradientn(colors = c("blue", "red"))
ggplot2::ggsave("unbiased_markers_heatmap.pdf",
                device = NULL,
                height = 8.5,
                width = 12,
                dpi = 500)

DoHeatmap(object = human_rett_cort_filt,features = top5_markers, size = 3)+ scale_fill_gradient2( low = rev(c('#d1e5f0','#67a9cf','#2166ac')), mid = "white", high = rev(c('#b2182b','#ef8a62','#fddbc7')), midpoint = 0, guide = "colourbar", aesthetics = "fill")

allen_markers <- c("GAD1", "LHX6", "ADARB2", "LAMP5", "VIP", "PVALB", "SST", "SLC17A7", " CUX2", "RORB", "THEMIS", "FEZF2", "CTGF", "AQP4", "PDGFRA", "OPALIN", "FYB")
Stacked_VlnPlot(seurat_object = human_rett_cort_filt, features = allen_markers)
DotPlot_scCustom(seurat_object = human_rett_cort_filt, features = allen_markers, x_lab_rotate = T, flip_axes = T, remove_axis_titles = FALSE)
ggsave("allen_markers_dotplot.pdf",
       device = NULL,
       height = 8.5,
       width = 12,
       dpi = 500)
DotPlot_scCustom(seurat_object = human_rett_cort_filt, features = top5_markers, x_lab_rotate = T, flip_axes = T, remove_axis_titles = FALSE)
ggsave("unbiased_markers_dotplot.pdf",
       device = NULL,
       height = 8.5,
       width = 12,
       dpi = 500)

DimPlot_scCustom(human_rett_cort_filt,  pt.size = 0.8, label = FALSE, figure_plot = TRUE)
ggsave("/Users/osman/Documents/GitHub/snRNA-seq-pipeline/scripts/07_human_cell_labeling/human_cortex_umap.pdf",
       device = NULL,
       height = 8.5,
       width = 12)
```



```{r human rett, fig.width=12, fig.height=10}

p1 <- FeaturePlot(hrett_cortex, features = "SST")
p2 <- FeaturePlot(hrett_cortex, features = "VIP")
p3 <- VlnPlot(hrett_cortex, features = c("FGFR3") , group.by = "predicted.cluster") + NoLegend()
p4 <- FeaturePlot(hrett_cortex, features = "predicted.cluster.score")
p1 + p2 + p3 + p4 + plot_layout(ncol=2)
```





As expected, query cells map to CD34+ celltypes which represent a subset of celltypes present in the reference. Rare cells that map to differentiated cell populations (i.e. CD4 memory), map with low prediction scores. Reference-mapping also harmonizes two separate runs.

```{r bm_plots, fig.width=10, fig.height=10}
p1 <- DimPlot_scCustom(seurat_object = hrett_cortex, group.by = "predicted.cluster", label = TRUE) + NoLegend()
p2 <- DimPlot(hrett_cortex, group.by = "orig.ident") 
sort(table(hrett_cortex$predicted.cluster), decreasing = TRUE)
p3 <- VlnPlot(hrett_cortex, features = "predicted.cluster.score") + NoLegend()
(p1 + p2) / p3

```

Lastly, we can visualize markers associated with lineage differentiation to verify our annotations including AVP (HSC), KLF1 (erythroid), MPO (myeloid), and VWF (platelet).

```{r bm_feature_plots, fig.width=12, fig.height=10}
# normalize before visualization
FeaturePlot_scCustom(hrett_cortex, features = c("VIP", "SST", "PVALB", "LAMP5"))
```
## Finally, lets save the filtered and normalized data
```{r}
save(human_rett_cort_filt, file="human_rett_cort_filt_labeled.RData")
save(list = ls(), file = "human_rett_cort_filt_labeled.RData")
```

```{r}
sessionInfo()
```
```{r test}
library(Seurat)
library(limma)

# Check if the cell_type column exists in the metadata
if (!("cell_type" %in% colnames(human_rettcort@meta.data))) {
  print("cell_type column not found in metadata. Skipping DEG analysis.")
} else {
  # Get the cell types
  cell_types <- human_rettcort@meta.data$cell_type

  # Check if there are any cell types
  if (length(cell_types) == 0) {
    print("No cell types found in cell_type column. Skipping DEG analysis.")
  } else {
    # Loop through each cell type
    for (cell_type in cell_types) {
      # Filter the cells for this cell type
      cell_type_filter <- human_rettcort@meta.data$cell_type == cell_type
      human_rettcort_cell_type <- human_rettcort[cell_type_filter, ]
      
      # Get the control and RTT samples
      control_samples <- human_rettcort_cell_type@meta.data$sample_type == "control"
      RTT_samples <- human_rettcort_cell_type@meta.data$sample_type == "RTT"
      
      # Get the control and RTT gene expression matrices
      control_mat <- human_rettcort_cell_type@assays$RNA@scale.data[, control_samples]
      RTT_mat <- human_rettcort_cell_type@assays$RNA@scale.data[, RTT_samples]
      
      # Get the control and RTT sample labels
      control_labels <- rep("control", sum(control_samples))
      RTT_labels <- rep("RTT", sum(RTT_samples))
      
      # Combine the control and RTT matrices and labels
      combined_mat <- cbind(control_mat, RTT_mat)
      sample_labels <- c(control_labels, RTT_labels)
      
      # Perform DEG analysis using Limma
      fit <- lmFit(combined_mat, model.matrix(~0 + sample_labels))
      contrast <- makeContrasts(RTT - control, levels = fit)
      fit2 <- contrasts.fit(fit, contrast)
      lrt <- lm.lik(fit2)
      pvals <- p.value(lrt)
      results <- topTable(fit2, adjust = "BH", number = Inf, sort.by = "none")
    }}}
```